### ISSUE TITLE
Cannot render preview with Julia Kernel

### ISSUE THREAD

**Issue by EvoArt** [quarto-dev/quarto-cli#2539]
### Bug description

Hi, I am hitting this error, trying to render a .qmd file, using Quarto in PowerShell (same with VSCode) on Windows 10. I can use the julia-1.7 kernel in jupyter lab, no probs. And I can render a .qmd using the python3 kernel in quarto
```
 Starting julia-1.7 kernel...Traceback (most recent call last):
   File "C:\Program Files\Quarto\share\jupyter\jupyter.py", line 21, in <module>
     from notebook import notebook_execute, RestartKernel
   File "C:\Program Files\Quarto\share\jupyter\notebook.py", line 17, in <module>
     from nbclient import NotebookClient
   File "C:\Users\arn203\.julia\conda\3\lib\site-packages\nbclient\__init__.py", line 6, in <module>
     from .client import NotebookClient, execute  # noqa: F401
   File "C:\Users\arn203\.julia\conda\3\lib\site-packages\nbclient\client.py", line 14, in <module>
     from jupyter_client import KernelManager
   File "C:\Users\arn203\.julia\conda\3\lib\site-packages\jupyter_client\__init__.py", line 6, in <module>
     from .asynchronous import AsyncKernelClient  # noqa
   File "C:\Users\arn203\.julia\conda\3\lib\site-packages\jupyter_client\asynchronous\__init__.py", line 1, in <module>
     from .client import AsyncKernelClient  # noqa
   File "C:\Users\arn203\.julia\conda\3\lib\site-packages\jupyter_client\asynchronous\client.py", line 6, in <module>
     from jupyter_client.channels import HBChannel
   File "C:\Users\arn203\.julia\conda\3\lib\site-packages\jupyter_client\channels.py", line 12, in <module>
     import zmq.asyncio
   File "C:\Users\arn203\.julia\conda\3\lib\site-packages\zmq\__init__.py", line 103, in <module>
     from zmq import backend
   File "C:\Users\arn203\.julia\conda\3\lib\site-packages\zmq\backend\__init__.py", line 31, in <module>
     raise original_error from None
   File "C:\Users\arn203\.julia\conda\3\lib\site-packages\zmq\backend\__init__.py", line 26, in <module>
     _ns = select_backend(first)
   File "C:\Users\arn203\.julia\conda\3\lib\site-packages\zmq\backend\select.py", line 31, in select_backend
     mod = import_module(name)
   File "C:\Users\arn203\.julia\conda\3\lib\importlib\__init__.py", line 127, in import_module
     return _bootstrap._gcd_import(name[level:], package, level)
   File "C:\Users\arn203\.julia\conda\3\lib\site-packages\zmq\backend\cython\__init__.py", line 6, in <module>
     from . import (
 ImportError: DLL load failed while importing _device: The specified module could not be found.
```
I'm pretty lost with this at this point! Any help would be much appreciated.

### Checklist

- [ ] Please include a minimal, fully reproducible example in a single .qmd file? Please provide the whole file rather than the snippet you believe is causing the issue.
- [X] Please [format your issue](https://quarto.org/bug-reports.html#formatting-make-githubs-markdown-work-for-us) so it is easier for us to read the bug report.
- [ ] Please document the RStudio IDE version you're running (if applicable), by providing the value displayed in the "About RStudio" main menu dialog?
- [X] Please document the operating system you're running. If on Linux, please provide the specific distribution.

**Comment by TS-CUBED** [comment]
This bug seems to be back with the new versions of Julia.

Anything newer than Julia 1.8.3 will not work.

Julia 1.8.4, 1.8.5, 1.9beta2, all fail with:

```
Starting julia-1.8 kernel...ERROR: 

Kernel died before replying to kernel_info
```

If I change the version number in `~/.local/share/jupyter/kernels/julia-1.8` from `1.8.4` to `1.8.3` it works.

All of the tested kernels work in all other instances where I use them: 

- Jupter notebook starts them properly
- Emacs Babel (`emacs-jupyter`) starts the Jupyter kernel as expected.

So, only quarto seems to have this problem.

**Comment by cscheid** [comment]
@TS-CUBED I'm trying to understand this better. I can't reproduce it on my side:

````
---
format: html
jupyter: julia-1.8
---

```{julia}
versioninfo()
```
````

<img width="783" alt="image" src="https://user-images.githubusercontent.com/285675/212134594-590972c8-e0fc-49f4-bec4-7741b37c9757.png">

The suggestion you provided changes information in `jupyter`:

> If I change the version number in `~/.local/share/jupyter/kernels/julia-1.8` from `1.8.4` to `1.8.3` it works.

I don't have a `~/.local/share/jupyter` folder in my setup, but I suspect this is causing your system trouble.

**Comment by TS-CUBED** [comment]
This problem has also been reported here:

https://discourse.julialang.org/t/quarto-julia-cli-kernel-died-before-replying-to-kernel-info/91416

https://github.com/quarto-dev/quarto-julia/issues/3

https://github.com/quarto-dev/quarto-cli/discussions/1652?sort=top

It seems to be a flaky issue, as in it does show up on a variety of systems (Mac, Win, Linux) with various installations of Jupyter (built-in, packaged, pip, conda).

The strange thing is that is sounds like a Python issue, but **only** occurs in quarto, the respective kernels work flawless in other clients that connect to a Jupyter server:

- Jupiter itself (notebook and lab)
- VSCode
- Emacs with Babel/Org-Mode

Is there a way to make quarto output useful debugging info?

**Comment by TS-CUBED** [comment]
Running your file I get:

```
Julia Version 1.8.3
Commit 0434deb161e (2022-11-14 20:14 UTC)
Platform Info:
  OS: Linux (x86_64-linux-gnu)
  CPU: 8 × Intel(R) Core(TM) i7-4770 CPU @ 3.40GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-13.0.1 (ORCJIT, haswell)
  Threads: 4 on 8 virtual cores
Environment:
  JULIA_NUM_THREADS = 4
```

Envious of your machine, btw ;-)

**Comment by TS-CUBED** [comment]
Kernel setup Jupyter:

```
{
  "display_name": "Julia 1.8.5",
  "argv": [
    "/home/thor/.julia/juliaup/julia-1.8.5+0.x64.linux.gnu/bin/julia",
    "-i",
    "--color=yes",
    "--project=@.",
    "/home/thor/.julia/packages/IJulia/6TIq1/src/kernel.jl",
    "{connection_file}"
  ],
  "language": "julia",
  "env": {},
  "interrupt_mode": "signal"
}
```

Results in:

```
Starting julia-1.8 kernel...ERROR: 

Kernel died before replying to kernel_info

Process quarto-preview-/home/thor/test.qmd exited abnormally with code 1
```

**Comment by prosoitos** [comment]
Having the same issue (also on Linux), so I can confirm that it is reproducible.

```sh
Julia Version 1.8.4
Commit 00177ebc4fc (2022-12-23 21:32 UTC)
Platform Info:
  OS: Linux (x86_64-linux-gnu)
  CPU: 16 × Intel(R) Core(TM) i7-10875H CPU @ 2.30GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-13.0.1 (ORCJIT, skylake)
  Threads: 1 on 16 virtual cores
```

**Comment by prosoitos** [comment]
In the Discourse post, the user cannot render. I can render (although it takes a surprisingly long time for the Julia kernel to start). What leads to the `Kernel died before replying to kernel_info` in my case is when I try to preview.

**Comment by TS-CUBED** [comment]
To rule out any hangup from my system config, I have just set up a new installation of Ubuntu 22.04.1 LTS

After the installation:

- Install jupyter:
```
sudo apt install python3-pip curl
python3 -m pip install jupyter
```

- Install Julia
```
curl -fsSL https://install.julialang.org | sh
```

- Install quarto-1.2.313-linux-amd64.deb from Quarto website.


This will install Julia 1.8.5.

In Julia REPL:

```
] add IJulia
] build IJulia
```

After this a `quarto render` of the file @cscheid  put up (MWE for a Julia .qmd) gives the `Kernel died before replying ...` error.

------------------------

```
julia add 1.8.3
julia +1.8.1
```
```
] add IJulia
] build IJulia
```

will install the 1.8.3 version as the `Julia-1.8` kernel in Jupyter, after which Quarto renders the file without a problem.

-----------------------

Also tested on the new Ubuntu  installation:

Kernel 1.8.5 and 1.9beta2 both work in the Jupyter notebook itself.

**Comment by prosoitos** [comment]
> If I change the version number in ~/.local/share/jupyter/kernels/julia-1.8 from 1.8.4 to 1.8.3 it works.

In my `~/.local/share/jupyter/kernels/julia-1.8`, I actually have `1.8.3`, which is puzzling since I am using `1.8.4`. So I must be using a different Jupyter installation.

Note that I did not install Python and Jupyter through Conda.jl.

**Comment by TS-CUBED** [comment]
> In the Discourse post, the user cannot render. I can render (although it takes a surprisingly long time for the Julia kernel to start). What leads to the Kernel died before replying to kernel_info in my case is when I try to preview.

Interesting. I can't render either. 

> In my ~/.local/share/jupyter/kernels/julia-1.8, I actually have 1.8.3, which is puzzling since I am using 1.8.4. So I must be using a different Jupyter installation.

It's quite easy to get out of sync between the Julia version you use in the REPL and the one that Jupyter uses. In order to udate the version Jupyter uses you need to do 

```
] add IJulia
] build IJulia
```

or it will continue to use the old one.

**Comment by prosoitos** [comment]
I can render, but it takes much longer than it should. So maybe the problem is a big hangup that my hardware is able to handle but yours can't?

What surprised me in the version of the kernel is that `versioninfo()` returns `1.8.4`. So I am using the new kernel.

**Comment by prosoitos** [comment]
(To clarify: I get `1.8.4` when I run `versioninfo()` in Quarto. So Quarto uses a Jupyter that uses the `1.8.4` kernel, but it doesn't seem to be the Jupyter that installed in `~/.local/share/jupyter/kernels/julia-1.8`).

I really thought I had a single Jupyter installation, so I am confused.

**Comment by prosoitos** [comment]
So ran `] build IJulia` and now I have `1.8.4` in `~/.local/share/jupyter/kernels/julia-1.8`, but I seem to be using it from somewhere else in Quarto anyway.

**Comment by prosoitos** [comment]
Ah, I found it: `/usr/share/jupyter/kernels/python3`.

**Comment by prosoitos** [comment]
Running `locate kernel.json`, the only other file I find is `/usr/share/jupyter/kernels/python3/kernel.json` which is for the ipython kernel. So I am really confused why Quarto was using a Julia kernel `1.8.4` before I had it installed...

**Comment by prosoitos** [comment]
Running `] build IJulia` and having the correct version in `~/.local/share/jupyter/kernels/julia-1.8` (`1.8.4` that `versioninfo()` was already returning when I had `1.8.3` there) seems to have magically solved the problem for me...

**Comment by schrimpf** [comment]
I am also encountering this issue.  I'm using Julia 1.8.5 and Quarto 1.2.313. `quarto preview` and `quarto render` both fail with:
```
ERROR: 

Kernel died before replying to kernel_info
```

The problem appears to be related to the kernel daemon. Rendering without the daemon works.
```bash
[test] $ quarto render test.qmd --execute-debug --no-execute-daemon
- Checking for existing kernel
- No existing kernel found

Starting julia-1.8 kernel...- Executing notebook with oneshot kernel
2023-01-13 11:55:08,672 - root - Level 25 - running notebook without keepalive
Done

Executing 'test.ipynb'
  Cell 1/1...Done

pandoc 
  to: html
  output-file: test.html
  standalone: true
  section-divs: true
  html-math-method: mathjax
  wrap: none
  default-image-extension: png
  
metadata
  document-css: false
  link-citations: true
  date-format: long
  lang: en
  title: Test
  author: Paul Schrimpf
  date: last-modified
  jupyter: julia-1.8
  
Output created: test.html
```

When using the daemon, I believe that it never even spawns a julia process. E.g.
```bash
[test] $ quarto render test.qmd --execute-debug --execute-daemon 300
- Connecting to kernel

Starting julia-log-1.8 kernel...2023-01-13 11:58:44,504 - root - Level 25 - starting notebook server subprocess
2023-01-13 11:58:44,504 - root - Level 25 - starting notebook server subprocess
2023-01-13 11:58:44,745 - root - Level 25 - running notebook server subprocess
2023-01-13 11:58:44,745 - root - Level 25 - creating notebook server (tcp: /run/user/1000/jt/a7993f57aa13b89de1b4)
2023-01-13 11:58:44,746 - root - Level 25 - notebook server bound to port 45535
2023-01-13 11:58:44,839 - root - Level 25 - handling server request
- Kernel connection successful
- Sending execute command to kernel
- Execute command sent, reading response
2023-01-13 11:58:44,840 - root - Level 25 - executing notebook
- Error response received
ERROR: 

Kernel died before replying to kernel_info
[test] $ ps -ef | grep python
paul        1086       1  0 Jan09 ?        00:03:17 /usr/bin/python /usr/bin/terminator
paul      295422       1  8 11:58 ?        00:00:00 /usr/bin/python3 /opt/quarto/share/jupyter/jupyter.
paul      295460  288134  0 11:58 pts/6    00:00:00 grep --color=auto python
[test] $ ps -ef | grep julia
paul      295463  288134  0 11:58 pts/6    00:00:00 grep --color=auto julia
```
So a python process running the jupyter server is spawned and left running, but there is no julia kernel left running. 

I've tried running this with all julia debug messages turned on (setting JULIA_DEBUG=all). For this, my kernel.json is
```json
{
  "display_name": "julia-log 1.8.5",
  "argv": [
    "/usr/bin/julia",
    "-i",
    "--color=yes",
    "/home/paul/.julia/packages/IJulia/6TIq1/src/kernel.jl",
    "{connection_file}"
  ],
  "language": "julia",
  "env": {
    "JULIA_DEBUG": "all"
  },
  "interrupt_mode": "signal"
}
```
with the daemon turned on and this kernel, there are no messages from julia. Either julia is never started, or it is crashing instantly before encountering any `@error`, `@info`, or `@debug` messages, which seems unlikely.

**Comment by mkitti** [comment]
Is there some common library dependency of both Jupyter/Python and Julia that is causing an issue?

**Comment by TS-CUBED** [comment]
> Is there some common library dependency of both Jupyter/Python and Julia that is causing an issue?

Possible, but I'd say unlikely, since the Kernel works fine in Jupyter itself and in other systems that use the jupyter backend to execute Julia code.

But in one of my earlier comments I tested that on a clean new Ubuntu installation, using the recommended installation from the Quarto docs. Which resulted in the same error.

**Comment by TS-CUBED** [comment]
I can get both preview and render working if I use the `--no-execute-daemon` option (or disable the daemon in the YAML frontmatter) once.

After quarto runs for the first time on a new kernel, I can then activate the daemon again and it works for subsequent runs.

So the problem seems to be with the first time initialisation of the daemon/connection to the daemon?

**Comment by binnisb** [comment]
I got the same running 1.9 beta. The `--no-execute-daemon` seems to work.

Strange since it worked for me yesterday in the same setup. I installed some new dependencies in the root env 1.9 and that might have done something, though I have no idea.

**Comment by pdeffebach** [comment]
On 1.8.4 the `--no-execute-daemon` option works for me. 

can you please let me know what this option does? How important is it?

**Comment by prosoitos** [comment]
https://quarto.org/docs/computations/julia.html#command-line

**Comment by cscheid** [comment]
Thanks for all your patience on this. It looks like we found the culprit. We don't understand why this crash is happening [deep inside Julia](https://github.com/quarto-dev/quarto-julia/issues/3#issuecomment-1367639678) but we have a fix that should be safe.

 We'll let the CI test suite run, and assuming things merge cleanly, we'll create a new prerelease build right after.

**Comment by urbainvaes** [comment]
Is it possible that, as a result of the option `no-execute-daemon`, the command `quarto preview` no longer works (it renders once in the beginning, but does not re-render on further changes).

**Comment by cscheid** [comment]
@urbainvaes Just for everyone else's clarity: no, that's not how the codebase works.

**Comment by urbainvaes** [comment]
@cscheid Thank you! For some reason the command now works, but it restarts the kernel at each change (and is very slow as a result). This is an issue I've had before, though, so it is probably not related to the `--no-execute-daemon` option.

**Comment by cscheid** [comment]
> but it restarts the kernel at each change (and is very slow as a result)

We'd love a repro and a new issue on this, thanks!

**Comment by benz0li** [comment]
@cderv I am not going to open a new issue as this [is resolved since v1.3.179 and] will be resolved with the official release of v1.3.

**Comment by cderv** [comment]
Great @benz0li Thanks for the feedback. Glad all is good;
